stl/inc/algorithm

```diff
@@ -3899,6 +3999,118 @@ _CONSTEXPR20 void pop_heap(_RanIt _First, _RanIt _Last) {
     _STD pop_heap(_First, _Last, less<>());
 }
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::pop_heap
+    template <class _It, class _Pr, class _Pj>
+    constexpr void _Pop_heap_hole_by_index(_It _First, iter_difference_t<_It> _Hole,
+        const iter_difference_t<_It> _Bottom, iter_value_t<_It>&& _Val, _Pr _Pred, _Pj _Proj) {
+        // percolate _Hole to _Bottom, then push _Val, using _Pred
+        _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+        _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
+        _STL_INTERNAL_CHECK(_Bottom != 0);
+        using _Diff      = iter_difference_t<_It>;
+        const _Diff _Top = _Hole;
+        _Diff _Idx       = _Hole;
+
+        // Check whether _Idx can have a child before calculating that child's index, since
+        // calculating the child's index can trigger integer overflows
+        const _Diff _Max_sequence_non_leaf = (_Bottom - 1) >> 1; // shift for codegen
+        while (_Idx < _Max_sequence_non_leaf) { // move _Hole down to larger child
+            _Idx            = 2 * _Idx + 2;
+            const auto _Mid = _First + _Idx;
+            if (_STD invoke(_Pred, _STD invoke(_Proj, *_Mid), _STD invoke(_Proj, *(_Mid - 1)))) {
+                --_Idx;
+            }
+            *(_First + _Hole) = _RANGES iter_move(_First + _Idx);
+            _Hole             = _Idx;
+        }
+
+        if (_Idx == _Max_sequence_non_leaf && _Bottom % 2 == 0) { // only child at bottom, move _Hole down to it

```

Comment:
So the optimizer can not divide by 2 but modulo is fine? Could you use  `& 1` if it is not?

Reply:
Modulo by powers of 2 doesn't have any weird pitfalls (https://godbolt.org/z/ikfJei).